tests/ctests/test_netplan_error.c

```diff
@@ -0,0 +1,66 @@
+#include <stdio.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+
+#include <cmocka.h>
+
+#include "../../include/netplan.h"
+#include "../../include/parse.h"
+
+#undef __USE_MISC
+#include "error.c"
+#include "names.c"
+#include "validation.c"
+#include "types.c"
+#include "util.c"
+#include "parse.c"
+
+void
+test_netplan_error_message(void** state)
+{
+    const gchar* message = "it failed";
+    char error_message[100] = {0};
+    GError *gerror = g_error_new(1, 2, "%s: error message", message);
+    netplan_error_message(gerror, error_message, sizeof(error_message) - 1);
+    assert_string_equal(error_message, "it failed: error message");
+    netplan_error_free(gerror);
+}
+
+void
+test_netplan_error_code(void** state)
+{
+    GError *gerror = g_error_new(1234, 5678, "%s: error message", "it failed");
+    uint64_t error_code = netplan_error_code(gerror);
+    GQuark domain = error_code >> 32;
+    gint error = error_code & ~0;
```

Comment:
question: Hmm... I wonder what the ` & ~0` is doing here? Does it have any effect? We're basically matching a (32 bit?) bitmask, like 0xFFFFFFFF, i.e. take the full integer. Should this rather be an (explicit) typecast?

Reply:
the idea was to mask the upper half of error code, but (looking at the generated code here) I suppose an implicit type casting is happening and my & ~0 is being optimized out in favor of an implicit cast. So yeah, an explicit cast would suffice.

btw, I assumed this was the semantics for netplan_error_code() based on the comment "uint64 to encode both domain and code" from the spec. Was that the idea? I feel like it would be better to return something like a `struct error` with 2 members...

Reply:
Good observation! But yes (having a 2nd look), that was the idea. We want to become GLib independent, so we don't really need the `domain` field, but also we don't want to drop that information for the time being. So masking/casting the relevant 32 bits will get the caller the correct error code now and in the future, without the need to change the API.